In [33]:
import pandas as pd 

In [34]:
df = pd.read_csv('train.csv')

In [35]:
df.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [36]:
df.tail()

PassengerId  Survived  Pclass                                      Name  \
886          887         0       2                     Montvila, Rev. Juozas   
887          888         1       1              Graham, Miss. Margaret Edith   
888          889         0       3  Johnston, Miss. Catherine Helen "Carrie"   
889          890         1       1                     Behr, Mr. Karl Howell   
890          891         0       3                       Dooley, Mr. Patrick   

        Sex   Age  SibSp  Parch      Ticket   Fare Cabin Embarked  
886    male  27.0      0      0      211536  13.00   NaN        S  
887  female  19.0      0      0      112053  30.00   B42        S  
888  female   NaN      1      2  W./C. 6607  23.45   NaN        S  
889    male  26.0      0      0      111369  30.00  C148        C  
890    male  32.0      0      0      370376   7.75   NaN        Q

## Lendo uma coluna específica arquivo csv

In [37]:
df= pd.read_csv('train.csv', usecols=['PassengerId', 'Survived','Pclass'])
df.head()

PassengerId  Survived  Pclass
0            1         0       3
1            2         1       1
2            3         1       3
3            4         1       1
4            5         0       3

In [38]:
df.shape

(891, 3)

## Getting some information about dataset with .describe() and .info()

### metado describe 
- Esse metado é usado para mostrar um sumário numerico do seu dataset
- calcula média, minimo, maximo, desvio padrão etcc
- poderemos ver se a valor ausente nas colunas

In [39]:
df = pd.read_csv('train.csv')

In [40]:
df.describe()

PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  714.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.699118    0.523008   
std     257.353842    0.486592    0.836071   14.526497    1.102743   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     223.500000    0.000000    2.000000   20.125000    0.000000   
50%     446.000000    0.000000    3.000000   28.000000    0.000000   
75%     668.500000    1.000000    3.000000   38.000000    1.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  
count  891.000000  891.000000  
mean     0.381594   32.204208  
std      0.806057   49.693429  
min      0.000000    0.000000  
25%      0.000000    7.910400  
50%      0.000000   14.454200  
75%      0.000000   31.000000  
max      6.000000  512.329200

#### Nós podemos ver que existe valor faltando na coluna age 714, sendo que deveria ter 891 

## métado .info 
- Esse métado printa as informações sobre seu dataframe incluindo index type e column type, valores non-null e mémoria usando

In [41]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


#### podemos usar tbm 

In [42]:
df.dtypes

PassengerId      int64
Survived         int64
Pclass           int64
Name            object
Sex             object
Age            float64
SibSp            int64
Parch            int64
Ticket          object
Fare           float64
Cabin           object
Embarked        object
dtype: object

# Classificando dados numéricos
- Existe uma coluna no DataFrame que repressenta o preço da passagem que o passageiro pagou. Vamos supor que queremos ver o valor mais baixo.

In [43]:
df.sort_values('Fare')

PassengerId  Survived  Pclass                                Name  \
271          272         1       3        Tornquist, Mr. William Henry   
597          598         0       3                 Johnson, Mr. Alfred   
302          303         0       3     Johnson, Mr. William Cahoone Jr   
633          634         0       1       Parr, Mr. William Henry Marsh   
277          278         0       2         Parkes, Mr. Francis "Frank"   
..           ...       ...     ...                                 ...   
438          439         0       1                   Fortune, Mr. Mark   
341          342         1       1      Fortune, Miss. Alice Elizabeth   
737          738         1       1              Lesurer, Mr. Gustave J   
258          259         1       1                    Ward, Miss. Anna   
679          680         1       1  Cardeza, Mr. Thomas Drake Martinez   

        Sex   Age  SibSp  Parch    Ticket      Fare        Cabin Embarked  
271    male  25.0      0      0      LINE    0.0000          NaN        S  
597    male  49.0      0      0      LINE    0.0000          NaN        S  
302    male  19.0      0      0      LINE    0.0000          NaN        S  
633    male   NaN      0      0    112052    0.0000          NaN        S  
277    male   NaN      0      0    239853    0.0000          NaN        S  
..      ...   ...    ...    ...       ...       ...          ...      ...  
438    male  64.0      1      4     19950  263.0000  C23 C25 C27        S  
341  female  24.0      3      2     19950  263.0000  C23 C25 C27        S  
737    male  35.0      0      0  PC 17755  512.3292         B101        C  
258  female  35.0      0      0  PC 17755  512.3292          NaN        C  
679    male  36.0      0      1  PC 17755  512.3292  B51 B53 B55        C  

[891 rows x 12 columns]

### ver os maiores valores de passagem 

In [44]:
df.sort_values("Fare", ascending = False).head(10)

PassengerId  Survived  Pclass                                   Name  \
258          259         1       1                       Ward, Miss. Anna   
737          738         1       1                 Lesurer, Mr. Gustave J   
679          680         1       1     Cardeza, Mr. Thomas Drake Martinez   
88            89         1       1             Fortune, Miss. Mabel Helen   
27            28         0       1         Fortune, Mr. Charles Alexander   
341          342         1       1         Fortune, Miss. Alice Elizabeth   
438          439         0       1                      Fortune, Mr. Mark   
311          312         1       1             Ryerson, Miss. Emily Borie   
742          743         1       1  Ryerson, Miss. Susan Parker "Suzette"   
118          119         0       1               Baxter, Mr. Quigg Edmond   

        Sex   Age  SibSp  Parch    Ticket      Fare            Cabin Embarked  
258  female  35.0      0      0  PC 17755  512.3292              NaN        C  
737    male  35.0      0      0  PC 17755  512.3292             B101        C  
679    male  36.0      0      1  PC 17755  512.3292      B51 B53 B55        C  
88   female  23.0      3      2     19950  263.0000      C23 C25 C27        S  
27     male  19.0      3      2     19950  263.0000      C23 C25 C27        S  
341  female  24.0      3      2     19950  263.0000      C23 C25 C27        S  
438    male  64.0      1      4     19950  263.0000      C23 C25 C27        S  
311  female  18.0      2      2  PC 17608  262.3750  B57 B59 B63 B66        C  
742  female  21.0      2      2  PC 17608  262.3750  B57 B59 B63 B66        C  
118    male  24.0      0      1  PC 17558  247.5208          B58 B60        C

# Classificando colunas com strings 

- Vamos ordenar a coluna 'Cabin. Existe diversos números faltando (NaN). Como podemos lidar com eles? Felizmente, há argumento chamado de posição que nos ajuda a definir uma posição para os valores de NaN no conjunto de dados

In [45]:
df.sort_values("Cabin", ascending = True, na_position ='last')

PassengerId  Survived  Pclass  \
583          584         0       1   
475          476         0       1   
556          557         1       1   
284          285         0       1   
599          600         1       1   
..           ...       ...     ...   
884          885         0       3   
885          886         0       3   
886          887         0       2   
888          889         0       3   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
583                                Ross, Mr. John Hugo    male  36.0      0   
475                        Clifford, Mr. George Quincy    male   NaN      0   
556  Duff Gordon, Lady. (Lucille Christiana Sutherl...  female  48.0      1   
284                         Smith, Mr. Richard William    male   NaN      0   
599       Duff Gordon, Sir. Cosmo Edmund ("Mr Morgan")    male  49.0      1   
..                                                 ...     ...   ...    ...   
884                             Sutehall, Mr. Henry Jr    male  25.0      0   
885               Rice, Mrs. William (Margaret Norton)  female  39.0      0   
886                              Montvila, Rev. Juozas    male  27.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch           Ticket     Fare Cabin Embarked  
583      0            13049  40.1250   A10        C  
475      0           110465  52.0000   A14        S  
556      0            11755  39.6000   A16        C  
284      0           113056  26.0000   A19        S  
599      0         PC 17485  56.9292   A20        C  
..     ...              ...      ...   ...      ...  
884      0  SOTON/OQ 392076   7.0500   NaN        S  
885      5           382652  29.1250   NaN        Q  
886      0           211536  13.0000   NaN        S  
888      2       W./C. 6607  23.4500   NaN        S  
890      0           370376   7.7500   NaN        Q  

[891 rows x 12 columns]

# Counting the occurences of variables

In [46]:
df['Sex'].value_counts()

male      577
female    314
Name: Sex, dtype: int64

# # Using .nunique() to count number of unique values that occur in dataset or in a column

In [47]:
df.nunique() #by typing this, we can see the counts of unique numbers in each column

PassengerId    891
Survived         2
Pclass           3
Name           891
Sex              2
Age             88
SibSp            7
Parch            7
Ticket         681
Fare           248
Cabin          147
Embarked         3
dtype: int64

In [48]:
df["Embarked"].nunique()

3

In [49]:
df[["Embarked", "Sex"]].nunique()

Embarked    3
Sex         2
dtype: int64

# Filtering

In [50]:
df["Embarked"] == "C" 

0      False
1       True
2      False
3      False
4      False
       ...  
886    False
887    False
888    False
889     True
890    False
Name: Embarked, Length: 891, dtype: bool

In [51]:
embarked_c_mask = df["Embarked"] == "C"
df[embarked_c_mask]

PassengerId  Survived  Pclass  \
1              2         1       1   
9             10         1       2   
19            20         1       3   
26            27         0       3   
30            31         0       1   
..           ...       ...     ...   
866          867         1       2   
874          875         1       2   
875          876         1       3   
879          880         1       1   
889          890         1       1   

                                                  Name     Sex   Age  SibSp  \
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
9                  Nasser, Mrs. Nicholas (Adele Achem)  female  14.0      1   
19                             Masselmani, Mrs. Fatima  female   NaN      0   
26                             Emir, Mr. Farred Chehab    male   NaN      0   
30                            Uruchurtu, Don. Manuel E    male  40.0      0   
..                                                 ...     ...   ...    ...   
866                       Duran y More, Miss. Asuncion  female  27.0      1   
874              Abelson, Mrs. Samuel (Hannah Wizosky)  female  28.0      1   
875                   Najib, Miss. Adele Kiamie "Jane"  female  15.0      0   
879      Potter, Mrs. Thomas Jr (Lily Alexenia Wilson)  female  56.0      0   
889                              Behr, Mr. Karl Howell    male  26.0      0   

     Parch         Ticket     Fare Cabin Embarked  
1        0       PC 17599  71.2833   C85        C  
9        0         237736  30.0708   NaN        C  
19       0           2649   7.2250   NaN        C  
26       0           2631   7.2250   NaN        C  
30       0       PC 17601  27.7208   NaN        C  
..     ...            ...      ...   ...      ...  
866      0  SC/PARIS 2149  13.8583   NaN        C  
874      0      P/PP 3381  24.0000   NaN        C  
875      0           2667   7.2250   NaN        C  
879      1          11767  83.1583   C50        C  
889      0         111369  30.0000  C148        C  

[168 rows x 12 columns]

# Finding the null values with .isnull()

In [52]:
null_mask = df["Cabin"].isnull()
df[null_mask]

PassengerId  Survived  Pclass                                      Name  \
0              1         0       3                   Braund, Mr. Owen Harris   
2              3         1       3                    Heikkinen, Miss. Laina   
4              5         0       3                  Allen, Mr. William Henry   
5              6         0       3                          Moran, Mr. James   
7              8         0       3            Palsson, Master. Gosta Leonard   
..           ...       ...     ...                                       ...   
884          885         0       3                    Sutehall, Mr. Henry Jr   
885          886         0       3      Rice, Mrs. William (Margaret Norton)   
886          887         0       2                     Montvila, Rev. Juozas   
888          889         0       3  Johnston, Miss. Catherine Helen "Carrie"   
890          891         0       3                       Dooley, Mr. Patrick   

        Sex   Age  SibSp  Parch            Ticket     Fare Cabin Embarked  
0      male  22.0      1      0         A/5 21171   7.2500   NaN        S  
2    female  26.0      0      0  STON/O2. 3101282   7.9250   NaN        S  
4      male  35.0      0      0            373450   8.0500   NaN        S  
5      male   NaN      0      0            330877   8.4583   NaN        Q  
7      male   2.0      3      1            349909  21.0750   NaN        S  
..      ...   ...    ...    ...               ...      ...   ...      ...  
884    male  25.0      0      0   SOTON/OQ 392076   7.0500   NaN        S  
885  female  39.0      0      5            382652  29.1250   NaN        Q  
886    male  27.0      0      0            211536  13.0000   NaN        S  
888  female   NaN      1      2        W./C. 6607  23.4500   NaN        S  
890    male  32.0      0      0            370376   7.7500   NaN        Q  

[687 rows x 12 columns]

In [53]:
df.isnull().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

# Lidando com valores ausentes 
- Existe muitas maneiras de lidar com valores faltante, nesse caso vamos 'ignorar a tupla' e 'preencher com a mediana'.Nós vamos ignorar a coluna 'Cabin' que tem 70% da coluna faltando.



#### Dropping a column 

In [54]:
df.drop(labels = ["Cabin"], axis=1).head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Embarked  
0      0         A/5 21171   7.2500        S  
1      0          PC 17599  71.2833        C  
2      0  STON/O2. 3101282   7.9250        S  
3      0            113803  53.1000        S  
4      0            373450   8.0500        S

## Preenchendo valores faltantes com .fillna()
- Para preencher valores faltante no dataframe, existe um métado chamado .fillna().
- Vamos assumir que nos temos muitos valores faltando na coluna e nos queremos preencher com zero.
- df["columnname"].fillna(0,inplace=True)
- Mas em vez de preencher com 'Unknown' preenchemos o valores faltante da coluna Age, vamos usar a mediana da coluna

In [57]:
df['Age'] = df['Age'].fillna(df['Age'].median())

### Checando se aida existe algum valor null na coluna Age

In [58]:
df['Age'].isnull().sum()

0